test/integration/scheduler/queue_test.go

```diff
@@ -0,0 +1,150 @@
+	testutils "k8s.io/kubernetes/test/integration/util"
+	imageutils "k8s.io/kubernetes/test/utils/image"
+)
+
+func TestServiceAffinityEnqueue(t *testing.T) {
+	profile := schedapi.KubeSchedulerProfile{
+		SchedulerName: v1.DefaultSchedulerName,
+		Plugins: &schedapi.Plugins{
+			PreFilter: schedapi.PluginSet{
+				Enabled: []schedapi.Plugin{
+					{Name: serviceaffinity.Name},
+				},
+			},
+			Filter: schedapi.PluginSet{
+				Enabled: []schedapi.Plugin{
+					{Name: serviceaffinity.Name},
+				},
+			},
+		},
+		PluginConfig: []schedapi.PluginConfig{
+			{
+				Name: serviceaffinity.Name,
+				Args: &schedapi.ServiceAffinityArgs{
+					AffinityLabels: []string{"hostname"},
+				},
+			},
+		},
+	}
+	// Use zero backoff seconds to bypass backoffQ.
+	testCtx := testutils.InitTestSchedulerWithOptions(
+		t,
+		testutils.InitTestMaster(t, "serviceaffinity-enqueue", nil),
+		nil,
+		scheduler.WithProfiles(profile),
+		scheduler.WithPodInitialBackoffSeconds(0),
+		scheduler.WithPodMaxBackoffSeconds(0),
+	)
+	testutils.SyncInformerFactory(testCtx)
+	// It's intended to not start the scheduler's queue, and hence to
+	// not start any flushing logic. We will pop and schedule the Pods manually later.
+	defer testutils.CleanupTest(t, testCtx)
+
+	cs, ns, ctx := testCtx.ClientSet, testCtx.NS.Name, testCtx.Ctx
+	// Create two Nodes.
+	for i := 1; i <= 2; i++ {
+		nodeName := fmt.Sprintf("node%d", i)
+		capacity := map[v1.ResourceName]string{v1.ResourcePods: "1"}
+		node := st.MakeNode().Name(nodeName).Label("hostname", nodeName).Capacity(capacity).Obj()
+		if _, err := cs.CoreV1().Nodes().Create(ctx, node, metav1.CreateOptions{}); err != nil {
+			t.Fatalf("Failed to create Node %q: %v", nodeName, err)
+		}
+	}
+
+	// Create a Service.
+	svc := &v1.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "svc",
+		},
+		Spec: v1.ServiceSpec{
+			Ports:    []v1.ServicePort{{Port: int32(80)}},
+			Selector: map[string]string{"foo": "bar"},
+		},
+	}
+	if _, err := cs.CoreV1().Services(ns).Create(ctx, svc, metav1.CreateOptions{}); err != nil {
+		t.Fatalf("Failed to create Service %q: %v", svc.Name, err)
+	}
+
+	// Create two Pods.
+	pause := imageutils.GetPauseImageName()
+	for i := 1; i <= 2; i++ {
+		podName := fmt.Sprintf("pod%d", i)
+		pod := st.MakePod().Namespace(ns).Name(podName).Label("foo", "bar").Container(pause).Obj()
+		// Make Pod1 an assigned Pod.
+		if i == 1 {
+			pod.Spec.NodeName = fmt.Sprintf("node%d", i)
+		}
+		if _, err := cs.CoreV1().Pods(ns).Create(ctx, pod, metav1.CreateOptions{}); err != nil {
+			t.Fatalf("Failed to create Pod %q: %v", pod.Name, err)
+		}
+	}
+
+	// Wait for pod2 to be present in the scheduling queue.
+	if err := wait.Poll(time.Millisecond*200, wait.ForeverTestTimeout, func() (bool, error) {
+		return len(testCtx.Scheduler.SchedulingQueue.PendingPods()) == 1, nil
+	}); err != nil {
+		t.Fatal(err)
+	}
+
+	// Pop Pod2 out. It should be unschedulable.
+	podInfo := testCtx.Scheduler.NextPod()
```

Comment:
should we expose scheduleOne instead?

Reply:
In very rare cases (like this one) we need that, so probably no.

Reply:
the same could be said about `NextPod` though.

Reply:
🤷